core/src/test/java/org/apache/iceberg/io/TestAppenderFactory.java

```diff
@@ -0,0 +1,265 @@
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.TableTestBase;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.deletes.EqualityDeleteWriter;
+import org.apache.iceberg.deletes.PositionDelete;
+import org.apache.iceberg.deletes.PositionDeleteWriter;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.Pair;
+import org.apache.iceberg.util.StructLikeSet;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public abstract class TestAppenderFactory<T> extends TableTestBase {
+  private static final int FORMAT_V2 = 2;
+
+  private final FileFormat format;
+  private final boolean partitioned;
+
+  private PartitionKey partition = null;
+
+  @Parameterized.Parameters(name = "FileFormat={0}, Partitioned={1}")
+  public static Object[] parameters() {
+    return new Object[][] {
+        new Object[] {"avro", false},
+        new Object[] {"avro", true},
+        new Object[] {"parquet", false},
+        new Object[] {"parquet", true}
+    };
+  }
+
+
+  public TestAppenderFactory(String fileFormat, boolean partitioned) {
+    super(FORMAT_V2);
+    this.format = FileFormat.valueOf(fileFormat.toUpperCase(Locale.ENGLISH));
+    this.partitioned = partitioned;
+  }
+
+  @Before
+  public void setupTable() throws Exception {
+    this.tableDir = temp.newFolder();
+    Assert.assertTrue(tableDir.delete()); // created by table create
+
+    this.metadataDir = new File(tableDir, "metadata");
+
+    if (partitioned) {
+      this.table = create(SCHEMA, SPEC);
+    } else {
+      this.table = create(SCHEMA, PartitionSpec.unpartitioned());
+    }
+    this.partition = createPartitionKey();
+
+    table.updateProperties()
+        .defaultFormat(format)
+        .commit();
+  }
+
+  protected abstract FileAppenderFactory<T> createAppenderFactory(List<Integer> equalityFieldIds,
+                                                                  Schema eqDeleteSchema,
+                                                                  Schema posDeleteRowSchema);
+
+  protected abstract T createRow(Integer id, String data);
+
+  protected abstract StructLikeSet expectedRowSet(Iterable<T> records) throws IOException;
+
+  protected abstract StructLikeSet actualRowSet(String... columns) throws IOException;
+
+  private OutputFileFactory createFileFactory() {
+    return new OutputFileFactory(table.spec(), format, table.locationProvider(), table.io(),
+        table.encryption(), 1, 1);
+  }
+
+  private PartitionKey createPartitionKey() {
+    if (table.spec().isUnpartitioned()) {
+      return null;
+    }
+
+    Record record = GenericRecord.create(table.spec().schema()).copy(ImmutableMap.of("data", "aaa"));
```

Comment:
I think this could just be `table.schema()`. No need to go through the spec to get the table schema.